ob1k-concurrent/src/main/java/com/outbrain/ob1k/concurrent/lazy/LazyComposableFuture.java

```diff
@@ -45,348 +45,355 @@
-    return new LazyComposableFuture<>(consumer -> executor.execute(() ->
-      consumer.consume(Try.apply(task::call))), delegateHandler ? executor : null);
-  }
-
-  public static <T> LazyComposableFuture<T> schedule(final Scheduler scheduler, final Callable<T> task, final long delay, final TimeUnit timeUnit) {
-    return new LazyComposableFuture<>(consumer -> scheduler.schedule(() ->
-      consumer.consume(Try.apply(task::call)), delay, timeUnit));
-  }
-
-  public static <T> LazyComposableFuture<T> collectFirst(final List<ComposableFuture<T>> futures) {
-    return new LazyComposableFuture<>(consumer -> {
-      final AtomicBoolean done = new AtomicBoolean();
-      for (final ComposableFuture<T> future : futures) {
-        future.consume(result -> {
-          if (done.compareAndSet(false, true)) {
-            consumer.consume(result);
-          }
+    private final Producer<T> producer;
+
+    private LazyComposableFuture(final Producer<T> producer) {
+        this.producer = producer;
+    }
+
+    public static <T> LazyComposableFuture<T> fromValue(final T value) {
+        return new LazyComposableFuture<>(consumer -> consumer.consume(Try.fromValue(value)));
+    }
+
+    public static <T> LazyComposableFuture<T> fromError(final Throwable error) {
+        return new LazyComposableFuture<>(consumer -> consumer.consume(Try.fromError(error)));
+    }
+
+    public static <T> ComposableFuture<T> build(final Producer<T> producer) {
+        return new LazyComposableFuture<>(producer);
+    }
+
+    public static <T> LazyComposableFuture<T> apply(final Supplier<T> supplier) {
+        return new LazyComposableFuture<>(consumer ->
+                consumer.consume(Try.apply(supplier::get)));
+    }
+
+    public static <T> LazyComposableFuture<T> submit(final Executor executor, final Callable<T> task, final boolean delegateHandler) {
+        LazyComposableFuture<T> res = new LazyComposableFuture<>(consumer -> {
+            executor.execute(() -> {
+                consumer.consume(Try.apply(task::call));
+            });
         });
-      }
-    });
-  }
-
-  public static <T> LazyComposableFuture<List<T>> collectAll(final List<ComposableFuture<T>> futures) {
-    return new LazyComposableFuture<>(consumer -> {
-      final AtomicInteger counter = new AtomicInteger(futures.size());
-      final AtomicBoolean errorTrigger = new AtomicBoolean(false);
-      final ConcurrentMap<Integer, Try<T>> results = new ConcurrentHashMap<>(futures.size());
-
-      int index = 0;
-      for (final ComposableFuture<T> future : futures) {
-        final int i = index++;
-        future.consume(result -> {
-          results.put(i, result);
-          if (result.isSuccess()) {
-            final int count = counter.decrementAndGet();
-            if (count == 0) {
-              consumer.consume(Try.fromValue(createResultList(results)));
+
+        return delegateHandler ? (LazyComposableFuture<T>) res.publishOn(executor) : res;
```

Comment:
since publishOn (in lazy future) affects only single handler in the chain, the extra #publishOn here is redundant - first handler when task value returns already runs in the executor thread, and this extra #publishOn only does extra context-switch and that's all.